File: plugin/scripts/codex-claude-compat.js
#!/usr/bin/env node
"use strict";

/*
 * Translate Reflexio's Claude CLI provider contract to `codex exec`.
 *
 * Reflexio shells out to CLAUDE_SMART_CLI_PATH as if it were Claude Code:
 *
 *   <path> -p --output-format stream-json --model <model> ...
 *
 * Under Codex, this small bridge preserves that executable contract while
 * routing the actual generation through the authenticated Codex CLI.
 */

const { spawnSync } = require("node:child_process");
const crypto = require("node:crypto");
const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");

const TIMEOUT_MS = 120_000;

function main(argv) {
  try {
    const { outputFormat, systemPrompt } = parseSupportedArgs(argv);
    const content = runCodex({
      prompt: fs.readFileSync(0, "utf8"),
      systemPrompt,
    });
    const payload =
      outputFormat === "stream-json"
        ? { type: "result", subtype: "success", result: content }
        : { result: content };
    process.stdout.write(`${JSON.stringify(payload)}\n`);
    return 0;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    process.stderr.write(`codex-claude-compat: ${message}\n`);
    return 1;
  }
}

function parseSupportedArgs(argv) {
  let outputFormat = "json";
  let systemPrompt = "";
  let idx = 0;
  while (idx < argv.length) {
    const arg = argv[idx];
    if (arg === "-p") {
      idx += 1;
    } else if (arg === "--output-format") {
      outputFormat = requireValue(argv, idx, arg);
      idx += 2;
    } else if (arg === "--model") {
      requireValue(argv, idx, arg);
      idx += 2;
    } else if (arg === "--verbose" || arg === "--include-partial-messages") {
      idx += 1;
    } else if (arg === "--append-system-prompt") {
      systemPrompt = requireValue(argv, idx, arg);
      idx += 2;
    } else {
      throw new Error(`unsupported Claude CLI argument: ${arg}`);
    }
  }
  if (outputFormat !== "json" && outputFormat !== "stream-json") {
    throw new Error(`unsupported --output-format: ${outputFormat}`);
  }
  return { outputFormat, systemPrompt };
}

function requireValue(argv, idx, name) {
  if (idx + 1 >= argv.length) {
    throw new Error(`${name} requires a value`);
  }
  return argv[idx + 1];
}

function runCodex({ prompt, systemPrompt }) {
  const codexPath = process.env.CLAUDE_SMART_CODEX_PATH || commandPath(codexNames());
  if (!codexPath) {
    throw new Error("codex CLI not found on PATH");
  }

  const outputPath = temporaryOutputPath();
  const args = [
    "exec",
    "--sandbox",
    "read-only",
    "--skip-git-repo-check",
    "--ephemeral",
    "--ignore-rules",
    "--output-last-message",
    outputPath,
    "-",
  ];
  const env = {
    ...process.env,
    CLAUDE_SMART_HOST: "codex",
    CLAUDE_SMART_INTERNAL: "1",
    CLAUDE_CODE_ENTRYPOINT: "optimizer",
  };

  try {
    const proc = spawnSync(codexPath, args, {
      input: codexPrompt({ prompt, systemPrompt }),
      encoding: "utf8",
      env,
      timeout: TIMEOUT_MS,
      windowsHide: true,
      shell: process.platform === "win32" && /\.(?:cmd|bat)$/i.test(codexPath),
    });
    if (proc.error) {
      if (proc.error.code === "ETIMEDOUT") {
        throw new Error(`codex CLI timed out after ${TIMEOUT_MS / 1000}s`);
      }
      throw proc.error;
    }
    if (proc.status !== 0) {
      const stderr = String(proc.stderr || "").trim().slice(0, 500);
      throw new Error(`codex CLI exited ${proc.status}: ${stderr}`);
    }
    const content = fs.readFileSync(outputPath, "utf8").trim();
    if (!content) {
      throw new Error("codex CLI returned empty output");
    }
    return content;
  } finally {
    try {
      fs.unlinkSync(outputPath);
    } catch {
      // Best effort cleanup only.
    }
  }
}

function codexNames() {
  return process.platform === "win32" ? ["codex.cmd", "codex.exe", "codex"] : ["codex"];
}

function commandPath(names) {
  const pathParts = (process.env.PATH || "").split(path.delimiter).filter(Boolean);
  for (const dir of pathParts) {
    for (const name of names) {
      const candidate = path.join(dir, name);
      if (fs.existsSync(candidate)) return candidate;
    }
  }
  return null;
}

function temporaryOutputPath() {
  const suffix = crypto.randomBytes(8).toString("hex");
  return path.join(os.tmpdir(), `claude-smart-codex-${process.pid}-${Date.now()}-${suffix}`);
}

function codexPrompt({ prompt, systemPrompt }) {
  if (!systemPrompt) return prompt;
  return `${systemPrompt}\n\n## Task\n${prompt}`;
}

process.exitCode = main(process.argv.slice(2));

File: plugin/scripts/codex-claude-compat.py
#!/usr/bin/env python3
"""Translate Reflexio's Claude CLI provider contract to ``codex exec``.

Reflexio's local provider currently shells out to ``CLAUDE_SMART_CLI_PATH`` as
if it were the Claude Code CLI:

    <path> -p --output-format json --model <model> [--append-system-prompt ...]

When claude-smart runs under Codex, this executable preserves that narrow
contract while routing the actual model call through ``codex exec``. The
stdout shape intentionally matches Claude Code's JSON output enough for
Reflexio's provider to read ``result``.
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path


_TIMEOUT_SECONDS = 120


def main(argv: list[str] | None = None) -> int:
    argv = list(sys.argv[1:] if argv is None else argv)
    try:
        system_prompt = _parse_supported_args(argv)
        content = _run_codex(
            prompt=sys.stdin.read(),
            system_prompt=system_prompt,
        )
    except Exception as exc:  # noqa: BLE001 - CLI bridge errors go to stderr.
        print(f"codex-claude-compat: {exc}", file=sys.stderr)
        return 1

    json.dump({"result": content}, sys.stdout, ensure_ascii=False)
    sys.stdout.write("\n")
    return 0


def _parse_supported_args(argv: list[str]) -> str:
    system_prompt = ""
    idx = 0
    while idx < len(argv):
        arg = argv[idx]
        if arg == "-p":
            idx += 1
        elif arg == "--output-format":
            idx += 2
        elif arg == "--model":
            idx += 2
        elif arg == "--append-system-prompt":
            if idx + 1 >= len(argv):
                raise ValueError("--append-system-prompt requires a value")
            system_prompt = argv[idx + 1]
            idx += 2
        else:
            raise ValueError(f"unsupported Claude CLI argument: {arg}")
    return system_prompt


def _run_codex(*, prompt: str, system_prompt: str) -> str:
    codex_path = os.environ.get("CLAUDE_SMART_CODEX_PATH") or shutil.which("codex")
    if not codex_path:
        raise FileNotFoundError("codex CLI not found on PATH")

    output_path = _temporary_output_path()
    cmd = [
        codex_path,
        "exec",
        "--sandbox",
        "read-only",
        "--skip-git-repo-check",
        "--ephemeral",
        "--ignore-rules",
        "--output-last-message",
        str(output_path),
        "-",
    ]

    env = os.environ.copy()
    env["CLAUDE_SMART_HOST"] = "codex"
    env["CLAUDE_SMART_INTERNAL"] = "1"
    env["CLAUDE_CODE_ENTRYPOINT"] = "optimizer"

    try:
        proc = subprocess.run(  # noqa: S603 - fixed command plus resolved executable.
            cmd,
            input=_codex_prompt(prompt=prompt, system_prompt=system_prompt),
            capture_output=True,
            text=True,
            timeout=_TIMEOUT_SECONDS,
            check=False,
            env=env,
        )
        if proc.returncode != 0:
            stderr = proc.stderr.strip()
            raise RuntimeError(f"codex CLI exited {proc.returncode}: {stderr[:500]}")
        content = output_path.read_text(encoding="utf-8").strip()
    except subprocess.TimeoutExpired as exc:
        raise TimeoutError(f"codex CLI timed out after {_TIMEOUT_SECONDS}s") from exc
    finally:
        try:
            output_path.unlink()
        except OSError:
            pass

    if not content:
        raise RuntimeError("codex CLI returned empty output")
    return content


def _temporary_output_path() -> Path:
    handle = tempfile.NamedTemporaryFile(prefix="claude-smart-codex-", delete=False)
    try:
        return Path(handle.name)
    finally:
        handle.close()


def _codex_prompt(*, prompt: str, system_prompt: str) -> str:
    if not system_prompt:
        return prompt
    return f"{system_prompt}\n\n## Task\n{prompt}"


if __name__ == "__main__":
    raise SystemExit(main())
